### PR TITLE
Adding variants support

### LINKED ISSUE #14669: Variants PR reviewed and checked in
Variants PR reviewed and checked in

### PR DESCRIPTION
**Pull request by dzearing**
This is an initial PR to support component variants using hooks; not the final form but wanted to get perf numbers earlier to see what needs tweaking. Fixes #14669

## Variants support

When building components, the component will come with a default stylesheet, piping in css variables (tokens) into the right place.

Each token may have a default value. Modifiers on the component may change the default values of those tokens. And, users will want to adjust those defaults through theming.

Variants are defined as such:

### 1. Create the variant hook 

Note this is created outside of the component scope.

```jsx
const useFooClasses = makeVariantClasses({
  name: 'Foo',
  prefix: '--foo', 
  styles: {
    root: { ... },
    icons: { ... },
  },
  variants: {
    root: { 
      /* baseline default token values */ 
      background: 'red'
    },
    primary: { 
      /* ...tokens which apply when the primary flag is set or the `variant` name on the component is "primary" */ 
      background: 'green'
    }
  }
});
```

### 2. The hook is used in a component

```jsx
const Foo = (props) => {
  const state = { ...props };

  useFooClasses (state);

  return (
    <div className={state.className} />
  );
```

### 3. The component can be used by the customer 

```jsx
<Foo /> // red foo
<Foo primary /> // green foo
<Foo variant="primary" /> // variant string prop can also be used
```

### 4. The component's variants can be themed

```jsx
<ThemeProvider theme={{
  components: {
     Foo: {
       variants: {
          primary: {
             background: 'purple'
          }
        }
      }
    }
  }}>
  <Foo/> // red button
  <Foo primary /> // purple button
</ThemeProvider>
```

### CODE REVIEW

File: packages/react-theme-provider/src/ThemeProvider.types.ts [hunk 1]
```diff
@@ -40,4 +41,6 @@ export interface ThemeProviderProps extends ComponentProps, React.HTMLAttributes
  */
 export type ThemeProviderState = Omit<ThemeProviderProps, 'theme'> & {
   theme: Theme;
+
+  customizerProps: ICustomizerContext;
```
Comment (xugao) [inline packages/react-theme-provider/src/ThemeProvider.types.ts:45]:
```suggestion
  customizerContext: ICustomizerContext;
```

File: packages/react-theme-provider/src/makeVariantClasses.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  // This function will only be called when styles have not been evaluated for this set for
+  // the particular theme/window/direction combo.
+  const styleFunction: (theme: Theme) => IStyle = (theme: Theme) => {
+    const themeVariants = name ? theme?.components?.[name]?.variants : undefined;
+
+    return [
+      callOrReturn(styles, theme),
+      processVariants(variants!, theme, name!, prefix!),
+      processVariants(themeVariants, theme, name!, prefix!),
```
Comment (xugao) [inline packages/react-theme-provider/src/makeVariantClasses.ts:82]: please remove/fix the usages of `!` here

File: packages/react-theme-provider/src/makeStyles.ts [hunk 3]
```diff
@@ -54,8 +54,11 @@ export function makeStyles<TStyleSet extends { [key: string]: IStyle }>(
     const win = useWindow();
 
     // Expected: theme and renderer are either always provided or never.
```
Comment (xugao) [inline packages/react-theme-provider/src/makeStyles.ts:56]: can remove comment

File: packages/react-button/src/components/Button/Button.types.ts [hunk 4]
```diff
@@ -87,61 +89,59 @@ export type ButtonProps = ComponentProps &
...
+      small?: string;
+      regular?: string;
+      large?: string;
+      larger?: string;
+      largest?: string;
+    };
+
+    pressed?: {
+      transform?: string;
+      transition?: string;
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.types.ts:137]: these fields look outdated

File: packages/react-button/src/components/Button/Button.test.tsx [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
+import { Button as BaseButton } from './Button';
 import * as renderer from 'react-test-renderer';
 import { mount, ReactWrapper } from 'enzyme';
 import { isConformant } from '../../common/isConformant';
+import { withThemeProvider } from '@fluentui/react-theme-provider';
 
-import { MergeStylesProvider } from '@fluentui/react-theme-provider';
-import { ButtonProps } from './Button.types';
+/** Use a ThemeProvider wrapper around the component to ensure styles show up in snapshots. */
+const Button = withThemeProvider(BaseButton);
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.test.tsx:9]: i think we should create helper in test-utils to handle this in `renderer.create`

File: packages/react-button/src/components/Button/Button.types.ts (hunk 4 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.types.ts:137]: hmmm. look right to me?

File: packages/react-theme-provider/src/makeVariantClasses.ts (hunk 2 above)
Comment (dzearing) [reply packages/react-theme-provider/src/makeVariantClasses.ts]: done!

File: packages/react-theme-provider/src/ThemeProvider.types.ts (hunk 1 above)
Comment (dzearing) [reply packages/react-theme-provider/src/ThemeProvider.types.ts:45]: done!

File: packages/react-button/src/components/Button/Button.types.ts (hunk 4 above)
Comment (xugao) [reply packages/react-button/src/components/Button/Button.types.ts:137]: if you compare this with useButtonClasses, tokens such as `whitespace` are missing

**Review by xugao — APPROVED** [review]

File: packages/react-theme-provider/src/makeStyles.ts (hunk 3 above)
Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts]: removed

**Pull request merged**